### PR TITLE
feat(portal): Accept hardware identifier fields and reset verification when they are changed

### PR DESCRIPTION
**Pull request by AndrewDryga**
If previous value is **null then** a new value is persisted and verification is **not removed**.
If previous value is **not null** then a new value is persisted and verification is **removed**.

Screenshot 2024-10-02 at 4 59 58 PM

### CODE REVIEW

File: elixir/apps/domain/lib/domain/clients/client/changeset.ex [hunk 1]
```diff
@@ -1,32 +1,131 @@
...
+        last_seen_remote_ip_location_lon: fragment("EXCLUDED.last_seen_remote_ip_location_lon"),
+        last_seen_version: fragment("EXCLUDED.last_seen_version"),
+        last_seen_at: fragment("EXCLUDED.last_seen_at"),
+        identity_id: fragment("EXCLUDED.identity_id"),
+        device_serial: fragment("EXCLUDED.device_serial"),
+        device_uuid: fragment("EXCLUDED.device_uuid"),
+        identifier_for_vendor: fragment("EXCLUDED.identifier_for_vendor"),
+        firebase_installation_id: fragment("EXCLUDED.firebase_installation_id"),
+        updated_at: fragment("NOW()"),
+        verified_at:
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/clients/client/changeset.ex:40]: Yes, this is ugly. I will try to make it better tomorrow but not 100% sure there is way to do it that won't be even worse 🤣

File: elixir/apps/domain/priv/repo/seeds.exs [hunk 2]
```diff
@@ -348,18 +348,21 @@ IO.puts("")
...
   Domain.Clients.upsert_client(
     %{
-      name: "FZ Admin iPhone",
+      name: "FZ Admin Laptop",
       external_id: Ecto.UUID.generate(),
       public_key: :crypto.strong_rand_bytes(32) |> Base.encode64(),
-      last_seen_user_agent: "iOS/12.7 (iPhone) connlib/0.7.412"
+      last_seen_user_agent: "Mac OS/14.5 (MacBook) connlib/0.7.412",
+      device_serial: "AAPL-SERIAL-#{Ecto.UUID.generate()}",
+      device_uuid: "AAPL-UUID-#{Ecto.UUID.generate()}"
```
Comment (jamilbk) [inline elixir/apps/domain/priv/repo/seeds.exs]:
```suggestion
      device_serial: "FVFHF246Q72Z",
      device_uuid: "#{Ecto.UUID.generate()}"
```

Minor, but these would be the realistic-ish values (to use for visual formatting / design)

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 3]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+      if is_nil(value) do
+        []
+      else
+        [{hardware_id_title(client, key), value}]
+      end
+    end)
+  end
+
+  defp hardware_id_title(%{last_seen_user_agent: "Mac OS/" <> _}, :device_serial),
+    do: {"Hardware's Serial", nil}
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
    do: {"Device Serial", nil}
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 4]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+
+  defp hardware_id_title(%{last_seen_user_agent: "Mac OS/" <> _}, :device_uuid),
+    do: {"Hardware's UUID", nil}
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
    do: {"Device UUID", nil}
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 5]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+    do:
+      {"Firebase installation ID",
+       "This value is reset if the Firezone application is reinstalled."}
+
+  defp hardware_id_title(_client, :device_serial),
+    do: {"Motherboard's Serial", nil}
+
+  defp hardware_id_title(_client, :device_uuid),
+    do: {"Motherboard's UUID", nil}
+
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
  defp hardware_id_title(_client, :device_serial),
    do: {"Device Serial", nil}

  defp hardware_id_title(_client, :device_uuid),
    do: {"Device UUID", nil}

```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 6]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+  defp hardware_id_title(_client, :identifier_for_vendor),
+    do: {"Identifier for vendor", nil}
+
+  defp hardware_id_title(_client, :firebase_installation_id),
+    do: {"Firebase installation ID", nil}
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
  defp hardware_id_title(_client, :firebase_installation_id),
    do: {"App installation ID", nil}
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 7]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+       "This value is reset if the Firezone application is reinstalled."}
+
+  defp hardware_id_title(_client, :device_serial),
+    do: {"Motherboard's Serial", nil}
+
+  defp hardware_id_title(_client, :device_uuid),
+    do: {"Motherboard's UUID", nil}
+
+  defp hardware_id_title(_client, :identifier_for_vendor),
+    do: {"Identifier for vendor", nil}
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
    do: {"App installation ID", nil}
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 8]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+  defp hardware_id_title(%{last_seen_user_agent: "Mac OS/" <> _}, :device_uuid),
+    do: {"Hardware's UUID", nil}
+
+  defp hardware_id_title(%{last_seen_user_agent: "iOS/" <> _}, :identifier_for_vendor),
+    do:
+      {"Identifier for vendor", "This value is reset if the Firezone application is reinstalled."}
+
+  defp hardware_id_title(%{last_seen_user_agent: "Android/" <> _}, :firebase_installation_id),
+    do:
+      {"Firebase installation ID",
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
      {"App installation ID",
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 9]
```diff
@@ -258,6 +303,49 @@ defmodule Web.Clients.Show do
...
+
+  defp hardware_id_title(%{last_seen_user_agent: "Mac OS/" <> _}, :device_serial),
+    do: {"Hardware's Serial", nil}
+
+  defp hardware_id_title(%{last_seen_user_agent: "Mac OS/" <> _}, :device_uuid),
+    do: {"Hardware's UUID", nil}
+
+  defp hardware_id_title(%{last_seen_user_agent: "iOS/" <> _}, :identifier_for_vendor),
+    do:
+      {"Identifier for vendor", "This value is reset if the Firezone application is reinstalled."}
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
      {"App installation ID", "This value is reset if the Firezone application is reinstalled."}
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 10]
```diff
@@ -190,11 +156,54 @@ defmodule Web.Clients.Show do
...
+                </:content>
+              </.popover>
+            </:label>
+            <:value><%= @client.id %></:value>
+          </.vertical_table_row>
+          <.vertical_table_row>
+            <:label>
+              <.popover>
+                <:target>
+                  Installation ID
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
                  File ID
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 11]
```diff
@@ -190,11 +156,54 @@ defmodule Web.Clients.Show do
...
+                  <.icon name="hero-question-mark-circle" class="w-3 h-3 mb-1 text-neutral-400" />
+                </:target>
+                <:content>
+                  Unique UUID generated and stored on the Client to identify it.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
                  Firezone-specific UUID generated and persisted to the device upon app installation.
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 12]
```diff
@@ -190,11 +156,54 @@ defmodule Web.Clients.Show do
...
-            <:value>
-              <.last_seen schema={@client} />
-            </:value>
+            <:label>
+              <.popover>
+                <:target>
+                  ID <.icon name="hero-question-mark-circle" class="w-3 h-3 mb-1 text-neutral-400" />
+                </:target>
+                <:content>
+                  Unique UUID assigned to this Client that can be used in the API to manage this Client.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
                 Database ID assigned to this Client that can be used to manage this Client via the REST API.
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 13]
```diff
@@ -190,11 +156,54 @@ defmodule Web.Clients.Show do
...
+        <.vertical_table id="posture">
           <.vertical_table_row>
-            <:label>Last seen remote IP</:label>
-            <:value>
-              <.last_seen schema={@client} />
-            </:value>
+            <:label>
+              <.popover>
+                <:target>
+                  ID <.icon name="hero-question-mark-circle" class="w-3 h-3 mb-1 text-neutral-400" />
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
                  Database ID <.icon name="hero-question-mark-circle" class="w-3 h-3 mb-1 text-neutral-400" />
```

File: elixir/apps/web/lib/web/live/clients/show.ex [hunk 14]
```diff
@@ -190,11 +156,54 @@ defmodule Web.Clients.Show do
               <.relative_datetime datetime={@client.last_seen_at} />
             </:value>
           </.vertical_table_row>
+        </.vertical_table>
+
+        <h2 class="mt-6 mb-4 text-xl leading-none tracking-tight text-neutral-900">
+          Posture
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/clients/show.ex]:
```suggestion
          Device Attributes
```

**Review by jamilbk — COMMENTED** [review]
Thanks for getting this implemented so quickly!

I think it might be good to normalize the attribute names, and then in our docs we can detail the platform-specific sources where these come from. That will make creating cross-platform posture-based conditions more straightforward.

Would it be possible to also move the client-specific details back up into the `Details` vertical table, and then just have the device IDs (`device_uuid`, `device_serial`) in the `Device Attributes` vertical table below?

That way we can add more device attributes later to that table (operating system version, disk encryption status, etc)

**Comment by AndrewDryga** [comment]
@jamilbk LMK if this is what you wanted:
Screenshot 2024-10-03 at 10 00 38 AM

- I renamed `Database ID` to `ID` to make it look less overloaded (it's on the Client section and has a tooltip anyway).
- I left the IP/location as part of Device attributes

**Review by jamilbk — APPROVED** [review]
LGTM

**Pull request merged**